### ISSUE TITLE
Support for PMW on ESP32?

### ISSUE THREAD

**Issue by vseven** [DanielOgorchock/ST_Anything#56]
Not really a issue but more of a FYI or possible improvement.  As I was playing with the idea of controlling analog RGB lights I came across issues with AnalogWrite on the ESP32.  I eventually figured it out and went on to try and create a RGB library based on your existing EX_Switch_Dim figuring it was the logical place to start (failing horrible but I'll keep trying).  In your EX_Switch_Dim library I saw this:

```
#if defined(ARDUINO_ARCH_ESP32)
	if (st::Executor::debug) {
		Serial.println(F("EX_Switch_Dim:: analogWrite not currently supported on ESP32!"));
	}
#else
		analogWrite(m_nPinPWM, map(m_nCurrentLevel, 0, 100, 0, 255));
#endif
```

I have a fix if you want to implement it but there is a little setup and it can get messy so not sure if you want to bother or wait until they possible support a standard AnalogWrite (which I'm starting to doubt).  But here is a short sketch that demonstrates using PWM on the ESP32 which you might be able to use since you are already checking for the existence of it:

```
uint8_t pwmPin = A4;
void setup()
{
// Attach the pin to a ledc "channel"
ledcAttachPin(pwmPin , 1);
// Set the channel frequency and resolution (12k hz and 8-bit here)
ledcSetup(1, 12000, 8);
}

void loop() {
//write a full on to channel 1
ledcWrite(1, 255);
delay(5000);
//write a full off to channel 1
ledcWrite(1, 0);
delay(5000);
}
```

There are 16 total channels for PWM so the above would work fine for one dimmer on a ESP32 but the second wouldn't work since the code would use the same channel.  So the only way around that would be to have the ledc channel be a passed in as a optional variable when using a ESP32.  (byte channelPWM maybe?)

I think that's what I'll have to do for a RGB device, pass in 3 pins and 3 optional channels for those pins.  I'm also aiming to make my RGB device usable on both a ESP8266 that supports standard analogWrite and the ESP32 which does it like above so if a ESP32 it would do the assigning channels thing.  At least that's the hope and use your code to determine if I'm on a ESP32 or not.

**Comment by vseven** [comment]
Looking over this the actual fix to use the EX_Switch_Dim would be just like the EX_RGB_Dim I made.  You'd need to pass in a PWM Channel on a ESP32 then when setting up the pin you'd actually attach the pin to the channel and setup the channel.   

Since you have a ESP32 Multiples file you can call it out in there and people using other chips (8266) wouldn't even know about the channel being needed.

**Comment by mlampkin** [comment]
Sorry - off topic but didn't want to reopen a previously closed thread.

You had mentioned trying to get get smartled strips working with an esp32 and having issues - in particular with timeouts / the watchdog kicking.  I don't have the code handy -and- only tried a few meters of strip -and- it was an old radio shack one ( hey $1 for 3 meters ) but... 

I was able to get it operational by firing off a new task for setting the leds using xTaskCreatePinnedToCore and set to core 0 and using hand rolled asm ( mostly nops ) to set my output pin / data high and low.  As long as I didn't exceed 3.2 seconds - the watchdog did NOT fire off / cause any problem.  I actually added extra delays just to test things / figure out if the docs where correct about how long I would have until the watchdog kicked.

Want me to try to dig it up / see if I can recreate it when I have time ?  I also have a simpler version for an uno without the task and just plain color changing ( no animation ).  I actually 99% certain that I KNOW where that code is - lol. :)
